module/ice-9/generic-hash-tables.scm

```diff
@@ -0,0 +1,854 @@
+         (,(first bindings) (ht-hash-function ,ht-var))
+	 (,(second bindings) (ht-associator ,ht-var)))
+     . ,body-forms))
+
+(define-syntax assert-mutable
+  (syntax-rules ()
+    ((assert-mutable ht)
+     (or (hash-table-mutable? ht)
+         (error "Hash table is not mutable" ht)))))
+
+
+;;;; Hashing
+
+;;; SRFI-125 and R6RS hash functions are supposed to accept only one
+;;; argument, but Guile standard hash tables needs two args.
+;;; Therefore, the hash functions inside the hash table always accepts
+;;; one (required) argument and at least one possible argument, which
+;;; must be a fixnum.
+
+;;; The largest fixnum is in `most-positive-fixnum' in module (guile),
+;;; though not documented anywhere but libguile/numbers.c.
+
+(define (caller-with-default-size hash-fn)
+  "Answer a function that makes `most-positive-fixnum' the default
+second argument to HASH-FN, a 2-arg procedure."
+  (lambda* (obj #:optional (size most-positive-fixnum))
+    (hash-fn obj size)))
+
+(define hash (caller-with-default-size (@ (guile) hash)))
+
+(define string-ci-hash string-hash-ci)
+
+(define hash-by-identity (caller-with-default-size hashq))
+(define hash-by-value    (caller-with-default-size hashv))
+
+(define (wrap-hash-function hash-function)
+  (lambda* (obj size)
+    (modulo (hash-function obj) size)))
+
+
+;;;; Reflective queries, construction, predicate
+
+(define (get-hash-functions equiv-function hash-function)
+  "Returns an internal and an external hash function."
+  (cond
+   (hash-function (cond
+                   ;; SRFI-69 should give HASH as default hash-function.
+                   ((or (eq? (@ (guile) hash) hash-function)
+                        (eq? hash hash-function))
+                    (values (@ (guile) hash) hash-function))
+                   ;; These procedures don't need to be wrapped.
+                   ((or (eq? string-hash hash-function)
+                        (eq? string-ci-hash hash-function))
+                    (values hash-function hash-function))
+                   ((or (eq? hashq hash-function)
+                        (eq? hash-by-identity hash-function))
+                    (values hashq hash-function))
+                   ((or (eq? hashv hash-function)
+                        (eq? hash-by-value hash-function))
+                    (values hashv hash-function))
+                   ;; Otherwise, wrap the given function.
+                   (else (values (wrap-hash-function hash-function) hash-function))))
+   ((eq? equal? equiv-function) (values (@ (guile) hash) hash))
+   ((eq? eq? equiv-function) (values hashq hash-by-identity))
+   ((eq? eqv? equiv-function) (values hashv hash-by-value))
+   ((eq? string=? equiv-function) (values string-hash string-hash))
+   ((eq? string-ci=? equiv-function) (values string-ci-hash string-ci-hash))
+   (else (error "A suitable hash function could not be determined" equiv-function))))
+
+(define (guile-ht-ctor weakness)
```

Comment:
Regarding `ht` I am not a great fan of shorthands, I would rather see `hashtable` instead. Also what is `ctor`?

Reply:
If the rest of guile code use `ht` keep it `ht`.

Reply:
Changed `ht` to `hash-table`, `ctor` to `constructor` and other variables to clearer names.